File: app/src/main/java/com/moisesborges/tvaddict/watching/WatchingShowsPresenter.java
package com.moisesborges.tvaddict.watching;

import android.support.annotation.NonNull;

import com.moisesborges.tvaddict.data.ShowsRepository;
import com.moisesborges.tvaddict.models.Show;
import com.moisesborges.tvaddict.mvp.BasePresenter;
import com.moisesborges.tvaddict.mvp.RxJavaConfig;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import timber.log.Timber;

/**
 * Created by Moisés on 22/05/2017.
 */

public class WatchingShowsPresenter extends BasePresenter<WatchingShowsView> {

    private ShowsRepository mShowsRepository;

    @Inject
    public WatchingShowsPresenter(@NonNull RxJavaConfig rxJavaConfig,
                                  @NonNull ShowsRepository showsRepository) {
        super(rxJavaConfig);
        mShowsRepository = showsRepository;
    }

    public void loadWatchingShows() {
        checkView();

        Disposable disposable = mShowsRepository.getWatchingShows()
                .compose(applySchedulersToSingle())
                .subscribe(shows -> {
                    if (shows.size() == 0) {
                        getView().displayEmptyListMessage();
                        return;
                    }

                    getView().displayWatchingShows(shows);
                });

        addDisposable(disposable);
    }
}

File: app/src/main/java/com/moisesborges/tvaddict/watching/WatchingShowsPresenter.kt
package com.moisesborges.tvaddict.watching

import com.moisesborges.tvaddict.data.ShowsRepository
import com.moisesborges.tvaddict.models.Show
import com.moisesborges.tvaddict.mvp.BasePresenter
import com.moisesborges.tvaddict.mvp.ContentObserverView
import com.moisesborges.tvaddict.mvp.RxJavaConfig
import com.moisesborges.tvaddict.mvp.asContentObserverView
import com.raizlabs.android.dbflow.runtime.FlowContentObserver
import com.raizlabs.android.dbflow.sql.language.SQLOperator
import com.raizlabs.android.dbflow.structure.BaseModel

import javax.inject.Inject

/**
 * Created by Moisés on 22/05/2017.
 */

class WatchingShowsPresenter
@Inject constructor(rxJavaConfig: RxJavaConfig,
                    private val showsRepository: ShowsRepository) : BasePresenter<WatchingShowsView>(rxJavaConfig) {

    var contentObserver: FlowContentObserver = FlowContentObserver()

    override fun bindView(view: WatchingShowsView) {
        super.bindView(view)
        view.asContentObserverView()?.registerContentObserver(contentObserver, Show::class.java)
        contentObserver.addModelChangeListener { clazz: Class<*>?, action: BaseModel.Action, arrayOfSQLOperators: Array<SQLOperator> -> loadWatchingShows() }
    }

    override fun unbindView() {
        view.asContentObserverView()?.unregisterContentObserver(contentObserver)
        super.unbindView()
    }

    fun loadWatchingShows() {
        checkView()

        val disposable = showsRepository.getSavedShows()
                .compose(applySchedulersToSingle<List<Show>>())
                .subscribe { shows ->
                    if (shows.isEmpty()) {
                        view.displayEmptyListMessage()
                    } else {
                        view.displayWatchingShows(shows)
                    }
                }

        addDisposable(disposable)
    }

    fun openShowDetails(show: Show) {
        checkView()

        view.navigateToShowDetails(show)
    }
}
